prow/bugzilla/client.go

```diff
@@ -214,37 +231,112 @@ func getRootForClone(c Client, bug *Bug) (*Bug, error) {
 
 // GetRootForClone returns the original bug.
 func (c *client) GetRootForClone(bug *Bug) (*Bug, error) {
-	return getRootForClone(c, bug)
+	bugCache := map[int]*Bug{}
+	return getRootForClone(c, bug, bugCache)
 }
 
 // GetAllClones returns all the clones of the bug including itself
 // Differs from GetClones as GetClones only gets the child clones which are one level lower
-func (c *client) GetAllClones(bug *Bug) ([]*Bug, error) {
-	return getAllClones(c, bug)
+func (c *client) GetAllClones(bugID int) ([]*Bug, *Bug, *Bug, error) {
+	bugCache := map[int]*Bug{}
+	return getAllClones(c, bugID, bugCache)
 }
-func getAllClones(c Client, bug *Bug) ([]*Bug, error) {
-	root, err := getRootForClone(c, bug)
+
+type clonesCacheWithMux struct {
+	cache map[int][]int
+	mux   sync.Mutex
+}
+
+var clonesCache = clonesCacheWithMux{cache: map[int][]int{}}
+
+func getAllClones(c Client, bugID int, intraCallCache map[int]*Bug) ([]*Bug, *Bug, *Bug, error) {
+
+	// Global clones cache should only be accessed after ensuring proper thread safety
+	clonesCache.mux.Lock()
+	clonesID, ok := clonesCache.cache[bugID]
+	clonesCache.mux.Unlock()
+	var bug *Bug
+	var getBugErr error
+	var wg sync.WaitGroup
+	cloneErrors := make([]error, len(clonesID))
+	clones := make([]*Bug, len(clonesID))
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// This error is being handled after the wg.Wait
+		bug, getBugErr = c.GetBug(bugID)
+	}()
+	if ok {
+		var intraCallCacheMux sync.Mutex
+		for index, cloneID := range clonesID {
+			wg.Add(1)
+			go func(index int, cloneID int) {
+				defer wg.Done()
+				var clone *Bug
+				intraCallCacheMux.Lock()
+				clone, ok = intraCallCache[cloneID]
+				intraCallCacheMux.Unlock()
+				if !ok {
+					var err error
+					clone, err = c.GetBug(cloneID)
+					if err != nil {
+						cloneErrors[index] = err
+						return
+					}
+				}
+				clones[index] = clone
```

Comment:
also not threadsafe. And it makes more sense to directly write into the cache rather than in a slice and then coipying that later in the func, because:
* Its easier to read
* You don't have to allocate the slice
* It might be re-used by one of the other goroutines and save that goroutine a call to Bugzilla

Reply:
Why would it not be threadsafe, I am writing stuff in different locations for a slice whose memory has already been allocated right? Using a slice like me would allow me to avoid using a lock in this scenario which would be needed if I were using a set/map.
1. Agreed, readability would improve if I were to assign it directly to the cache
3. I don't think so as all the clones in the list would be distinct.

Reply:
Okay the pre-allocation makes it threadsafe but there is no verification for this (someone might change this in the future) and its harder to read and a data structure we don't really need, we will copy it over anyways